File: Python/space/triangulation_view_test.py
from ..datastructures.tree_view import MetaRootView, NodeView
from .triangulation import InitialTriangulation
from .triangulation_view import TriangulationView


def test_vertex_subtree():
    T = InitialTriangulation.unit_square()
    elem_meta_root = T.elem_meta_root
    elem_meta_root.uniform_refine(6)

    vertex_subtree = MetaRootView.from_metaroot(T.vertex_meta_root)
    T_view = TriangulationView(vertex_subtree)
    assert len(T_view.elements) == 2

    # Create a subtree with only vertices lying below the diagonal.
    vertex_subtree = MetaRootView.from_metaroot_deep(
        T.vertex_meta_root,
        call_filter=lambda vertex: vertex.x + vertex.y <= 1)
    assert len(vertex_subtree.bfs()) < len(T.vertex_meta_root.bfs())
    T_view = TriangulationView(vertex_subtree)
    assert len(T_view.elements) < len(T.elem_meta_root.bfs())

    # Check that the history object contains exactly non-root vertices
    assert len(T_view.history) == len(vertex_subtree.bfs()) - len(
        vertex_subtree.roots)

    # Check that we do not have duplicates
    vertices_subtree = set(v.node for v in vertex_subtree.bfs())
    assert len(vertex_subtree.bfs()) == len(vertices_subtree)

    # Check all nodes necessary for the elem subtree are inside the vertices_subtree
    for elem in T_view.elements:
        for vtx in elem.node.vertices:
            assert vtx in vertices_subtree

    # And the other way around.
    elements_subtree = set(e.node for e in T_view.elements)
    assert len(T_view.elements) == len(elements_subtree)
    for vertex in vertex_subtree.bfs():
        for elem in vertex.node.patch:
            assert elem in elements_subtree

File: C++/space/triangulation_view_test.cpp

#include "space/triangulation_view.hpp"

#include <cmath>
#include <map>
#include <set>

#include "datastructures/multi_tree_view.hpp"
#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "space/initial_triangulation.hpp"
#include "space/triangulation.hpp"

using namespace space;
using namespace datastructures;
using ::testing::ElementsAre;

constexpr int max_level = 6;

TEST(TriangulationView, UniformRefine) {
  auto T = InitialTriangulation::UnitSquare();
  T.elem_tree.UniformRefine(max_level);

  for (int level = 0; level <= max_level; ++level) {
    auto vertex_view = TreeView<Vertex>(T.vertex_meta_root);
    vertex_view.UniformRefine(level);

    std::vector<Vertex *> vertices;
    for (auto vtx : vertex_view.Bfs()) vertices.emplace_back(vtx->node());

    // Now create the corresponding element tree
    TriangulationView triang_view(std::move(vertices));

    // Lets see if this actually gives some fruitful results!
    ASSERT_EQ(triang_view.element_leaves().size(), pow(2, level + 1));

    for (const auto &[elem, _] : triang_view.element_leaves()) {
      ASSERT_TRUE(elem->level() == level);
    }
  }
}

TEST(TriangulationView, VertexSubTree) {
  auto T = InitialTriangulation::UnitSquare();
  T.elem_tree.UniformRefine(max_level);

  // Create a subtree with only vertices lying below the diagonal.
  auto vertex_subtree = TreeView<Vertex>(T.vertex_meta_root);
  vertex_subtree.DeepRefine(/* call_filter */ [](const auto &vertex) {
    return vertex->level() == 0 || (vertex->x + vertex->y <= 1.0);
  });
  ASSERT_TRUE(vertex_subtree.Bfs().size() < T.vertex_meta_root->Bfs().size());

  auto T_view = TriangulationView(vertex_subtree);

  // Check there are no duplicates.
  std::set<Vertex *> vertices_subtree;
  for (auto vertex : T_view.vertices()) {
    vertices_subtree.insert(vertex);
  }
  ASSERT_EQ(vertices_subtree.size(), T_view.vertices().size());
  // Check all nodes necessary for the elem subtree are
  // inside the vertices_subtree.
  for (const auto &[elem, _] : T_view.element_leaves()) {
    for (auto &vtx : elem->vertices()) {
      ASSERT_TRUE(vertices_subtree.count(vtx));
    }
  }
}
